### DISCUSSION TITLE
Piper tts not functioning properly on RPI5

### DISCUSSION THREAD

**Discussion by LizardKing2212** [rhasspy/piper#359] [Q&A] [answered]
Hello all! I have run into an issue with using Piper tts on my pi 5. I have followed directions for installation and when running a default model that piper provides it generates choppy and skips over words. I have been working on this issue for days and still cannot solve it! The end goal is to use this custom model: https://huggingface.co/poisson-fish/piper-vasco/tree/main/onnx

but I cannot seem to figure this out. I am new to this so maybe that's why? Any and all help is greatly appreciated. 
The output does generate and I can hear some words, but like I previously mentioned, they are super jumbled and less than half of the input text is said.

If you take a look at the link above, the huggingface model is .onnx and also has a onnx.json config file. I even employed chat gpt to help with some of the errors but even that did not work.

-Lizardking

**Comment by MaxTheTurtle0** [comment]
I am facing the same issue

**Comment by DaveXanatos** [comment]
Lizardkingm MaxTheTurtle0 - Also having issues on Pi5 - only mine says it can't find the .onnx & json files, despite their existence in the same directory as piper.  Can anyone offer any suggestions?

**Reply by LizardKing2212** [reply]
Did you specifically provide the entire path to the model? That sometimes works

**Comment by LizardKing2212** [comment]
After quite a bit of work, and a lot of research, I determined that there is a problem when converting the generated speech into an audio format. I still do not know why this is or how to fix it, but what worked for me is to just use the direct audio output option for piper through the terminal. The command is:

echo 'This is what is spoken' | \
  ./piper --model vasco.onnx --config onxx_vasco.onnx.json (I think that's the config file) --output-raw | \
  aplay -r 22050 -f S16_LE -t raw -

Aplay comes installed with the system, but just ensure you have it installed and updated. Also good to run a quick sudo apt-get update && upgrade

This did fix the problem for me, but let me know if you both run into any more errors. Again, I still have no idea why saving the audio file won't work properly.

**Comment by DaveXanatos** [comment, marked as answer]
I did get it working:

```
wget https://github.com/rhasspy/piper/releases/download/v1.2.0/piper_arm64.tar.gz
tar -xvzf piper_arm64.tar.gz
```

Will now have subfolder, cd to it
Get the voice model & json

```
    wget https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/en/en_US/amy/medium/en_US-amy-medium.onnx
    wget https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/en/en_US/amy/medium/en_US-amy-medium.onnx.json
```

You should have two files -  en_US-amy-medium.onnx  and en_US-amy-medium.onnx.json - in same dir as piper

In same dir type echo 'Welcome to the world of speech synthesis!' | ./piper --model en_US-amy-medium.onnx --output file welcome.wav

```
pi@LangDev:~/Desktop/Piper/piper $ ls
en_US-amy-medium.onnx       libonnxruntime.so.1.14.1
en_US-amy-medium.onnx.json  libpiper_phonemize.so
espeak-ng-data              libpiper_phonemize.so.1
libespeak-ng.so             libpiper_phonemize.so.1.1.0
libespeak-ng.so.1           libtashkeel_model.ort
libespeak-ng.so.1.1.51      piper
libonnxruntime.so
pi@LangDev:~/Desktop/Piper/piper $ echo 'Welcome to the world of speech synthesis!' | ./piper --model en_US-amy-medium.onnx --output file welcome.wav


pi@LangDev:echo 'Well holy shit, I've got a bunch of really good new voices, and they're free!  This will be the new base set of voices for all Raspberry Pie Voice-Enabled applications going forward.  Welcome to the new world of voice tech!' | ./piper --model en_US-l2arctic-medium.onnx --output file "welcome.wav"
```

But the file gets saved as - in this case - "1714958049462550118.wav" instead of welcome.wav

I used the same pipe to aplay as you and got better playback times anyway, so not an issue.

Thanks for your help

**Reply by LizardKing2212** [reply]
Oh my bad, I provided the config and model that I was using instead of the general ones haha! Glad you got it working though, for what I'm doing I decided I don't really need to use the saving function and opted for aplay which as you mentioned does have better results.

**Reply by DaveXanatos** [reply]
Yes, unless you're using the audio in a video or something, real-time voice output is all that is needed.  In my case, it will be speaking the output of an LLM like GPT-4 - but hopefully a locally running version like llama...

**Reply by DaveXanatos** [reply]
In case anyone can use it, here's a python script that calls piper-tts.  Put it in the same directory as the piper executable.  It will speak whatever you type at the prompt. Do NOT use quotes (single or double) in your text inputs... I have to write some code to escape those lol

Change to whatever voice model you like...  I downloaded them all, so :)

```
#PiperSpeechPi5.py
#Updated 2024-05-06, David Xanatos

import os
import re

pronunciationDict = {'teh':'the'}  #Because I am genetically incapable of typing "the" without stopping my entire brain first
def adjustResponse(response):     # Use to adjust spellings in input string to create better speech output.  Add as many as you like...
    for key, value in pronunciationDict.items():
        if key in response or key.lower() in response:
            response = re.sub(key, value, response, flags=re.I)
    return response

def main():
    while True:
        SpeakText = input("(No Quotes for now...)> ")
        SpeakText = adjustResponse(SpeakText)
        print("Text = " + SpeakText + "\n")
        os.system("echo '" + SpeakText + "' | ./piper --model en_US-hfc_female-medium.onnx --output-raw | aplay -r 22050 -f S16_LE -t raw -")
        if SpeakText == "quit":
            break
if __name__ == "__main__":
    main()

```

**Reply by kyleti** [reply]
I get the following using a Raspberry PI 4 
bash: ./piper: No such file or directory
I'm a Linux newbie so I might be making a dump mistake.

**Reply by LizardKing2212** [reply]
> I get the following using a Raspberry PI 4 
> 
> bash: ./piper: No such file or directory
> 
> I'm a Linux newbie so I might be making a dump mistake.

How did you install piper? Through pip? Or through git?

**Reply by simobov87** [reply]
> I get the following using a Raspberry PI 4 bash: ./piper: No such file or directory I'm a Linux newbie so I might be making a dump mistake.

I'm getting the same result. 
I've just extracted piper into Downloads folder, then downloaded a voice, but still get this error.
I've tried versions 1.2.0 and also 2023.11.14-2

how to solve the issue?
I've also tried to install espeak manually from apt-get but nothing changed
